syntax = "proto3";

package customer;

service CustomerService {
  rpc CreateBuyer(CreateBuyerRequest) returns (CreateBuyerResponse);
  rpc ServerCreateBuyer(ServerCreateBuyerRequest) returns (ServerCreateBuyerResponse);

  rpc CreateCart(CreateCartRequest) returns (CreateCartResponse);
  rpc ServerCreateCart(CreateCartRequest) returns (CreateCartResponse);

  rpc DeleteCartByBuyerId(DeleteCartByBuyerIdRequest) returns (DeleteCartByBuyerIdResponse);
  rpc ServerDeleteCartByBuyerId(DeleteCartByBuyerIdRequest) returns (DeleteCartByBuyerIdResponse);


  rpc UpdateCartItemQuantity(UpdateCartItemQuantityRequest) returns (UpdateCartItemQuantityResponse);
  rpc ServerUpdateCartItemQuantity(UpdateCartItemQuantityRequest) returns (UpdateCartItemQuantityResponse);

  rpc RemoveCartItem(RemoveCartItemRequest) returns (RemoveCartItemResponse);
  rpc ServerRemoveCartItem(RemoveCartItemRequest) returns (RemoveCartItemResponse);

  rpc GetCartItem(GetCartItemRequest) returns (GetCartItemResponse);
  rpc GetBuyerCartItems(GetBuyerCartItemsRequest) returns (GetBuyerCartItemsResponse);
  rpc GetBuyerPurchase(GetBuyerPurchaseRequest) returns(GetBuyerPurchaseResponse);

  rpc GetBuyerId(GetBuyerIdRequest) returns (GetBuyerIdResponse);
}

message GetBuyerPurchaseRequest{
  int32 buyer_id=1;
}

message GetBuyerPurchaseResponse {
  repeated Item item_list=1;
}

message GetBuyerCartItemsRequest {
  int32 buyer_id=1;
}

message GetBuyerCartItemsResponse{
  repeated Item items = 1;
}

message RemoveCartItemRequest{
  int32 item_id=1;
  int32 buyer_id=2;
}

message RemoveCartItemResponse{
  bool success = 1;

}

message UpdateCartItemQuantityRequest{
  int32 item_id = 1;
  int32 buyer_id = 2;
  int32 quantity = 3;
}

message UpdateCartItemQuantityResponse{
  bool success = 1;
}

message GetCartItemRequest {
  int32 item_id = 1;
  int32 buyer_id  = 2;
}

message GetCartItemResponse{
  Item item = 1;
}

message DeleteCartByBuyerIdRequest {
  int32 buyer_id = 1;
}

message DeleteCartByBuyerIdResponse {
  bool success=1;
}

message CreateCartRequest {
  string item_name = 1;
  int32  item_id = 2;
  int32 buyer_id = 3;
  int32 quantity = 4;
  double price = 5;
}

message CreateCartResponse {
  bool success=1;
}

message GetBuyerIdRequest {
  string username = 1;
  string password = 2;
}

message GetBuyerIdResponse {
  int32 buyer_id = 1;
}

message  CreateBuyerRequest{
  string username = 1;
  string password = 2;
}

message CreateBuyerResponse{
  bool success = 1;
}

message  ServerCreateBuyerRequest{
  string username = 1;
  string password = 2;
}

message ServerCreateBuyerResponse{
  bool success = 1;
}




// Item details
message Item {
  int32 id = 1;
  int32 buyer_id = 2;
  int32 item_id = 3;
  int32 quantity = 4;
  double price = 5;
  string name=6;
}

